connect/runtime/src/test/java/org/apache/kafka/connect/storage/OffsetStorageWriterTest.java

```diff
@@ -192,6 +212,251 @@ public void testCancelAfterAwaitFlush() throws Exception {
+                () -> LoggingContext.forConnector("source-connector"),
+                connectorStore,
+                "offsets-topic",
+                mock(TopicAdmin.class));
+
+        OffsetStorageWriter offsetStorageWriter = new OffsetStorageWriter(offsetBackingStore, NAMESPACE, keyConverter, valueConverter);
+
+        offsetStorageWriter.offset(OFFSET_KEY, OFFSET_VALUE);
+        assertTrue(offsetStorageWriter.beginFlush(1000L, TimeUnit.MILLISECONDS));
+        Future<Void> flushFuture = offsetStorageWriter.doFlush((error, result) -> {
+            assertEquals(PRODUCE_EXCEPTION, error);
+            assertNull(result);
+        });
+        assertThrows(ExecutionException.class, () -> flushFuture.get(1000L, TimeUnit.MILLISECONDS));
+    }
+
+    @SuppressWarnings("unchecked")
+    private KafkaOffsetBackingStore setupOffsetBackingStoreWithProducer(String topic, boolean throwingProducer) {
+        KafkaOffsetBackingStore offsetBackingStore = new KafkaOffsetBackingStore(() -> mock(TopicAdmin.class), () -> "connect",  mock(Converter.class));
+        MockConsumer<byte[], byte[]> consumer = new MockConsumer<>(OffsetResetStrategy.EARLIEST);
+        Node noNode = Node.noNode();
+        Node[] nodes = new Node[]{noNode};
+        consumer.updatePartitions(topic, Collections.singletonList(new PartitionInfo(topic, 0, noNode, nodes, nodes)));
+        KafkaBasedLog<byte[], byte[]> kafkaBasedLog = new KafkaBasedLog<byte[], byte[]>(
+                topic, new HashMap<>(), new HashMap<>(),
+                () -> mock(TopicAdmin.class), mock(Callback.class), new MockTime(), null) {
+            @Override
+            protected Producer<byte[], byte[]> createProducer() {
+                return createMockProducer(throwingProducer);
+            }
+
+            @Override
+            protected Consumer<byte[], byte[]> createConsumer() {
+                return consumer;
+            }
+        };
+        kafkaBasedLog.start();
+        offsetBackingStore.offsetLog = kafkaBasedLog;
+        return offsetBackingStore;
+    }
+
+
+    private Producer<byte[], byte[]> createMockProducer(boolean throwingProducer) {
+        if (throwingProducer) {
+            return new MockProducer<byte[], byte[]>() {
+                @Override
+                public synchronized Future<RecordMetadata> send(final ProducerRecord<byte[], byte[]> record, final org.apache.kafka.clients.producer.Callback callback) {
+                    callback.onCompletion(null, PRODUCE_EXCEPTION);
+                    return null;
+                }
+            };
+        }
+        return new MockProducer<byte[], byte[]>() {
+            @Override
+            public synchronized Future<RecordMetadata> send(final ProducerRecord<byte[], byte[]> record, final org.apache.kafka.clients.producer.Callback callback) {
+                callback.onCompletion(null, null);
+                return null;
+            }
+        };
```

Comment:
nit: can be simplified
```suggestion
        return new MockProducer<byte[], byte[]>() {
            @Override
            public synchronized Future<RecordMetadata> send(final ProducerRecord<byte[], byte[]> record, final org.apache.kafka.clients.producer.Callback callback) {
                if (throwingProducer) {
                    callback.onCompletion(null, PRODUCE_EXCEPTION);
                } else {
                    callback.onCompletion(null, null);
                }
                return null;
            }
        };
```